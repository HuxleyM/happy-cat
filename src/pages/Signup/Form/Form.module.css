@import "../../../variables.css";

.navBar > ul > li {
  display: inline-block;
  padding: 5px;
  margin-right: 10px;
  border-bottom: 3pt solid var(--colour-grey);
}

.header {
  font-size: 28px;
  font-weight: 700;
  margin-right: 100px;
}

.headerWrapper {
  display: flex;
  justify-content: space-between;
  flex-wrap: wrap;
  margin-bottom: 24px;
}

.addBreathingSpace {
  padding-bottom: 28px;
  border-bottom: 3pt solid var(--colour-grey);
}
