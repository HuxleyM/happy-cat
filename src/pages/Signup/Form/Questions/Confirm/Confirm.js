import React from "react";


function Confirm() {

  return (<h2>Heyy</h2>)
}

export default Confirm;
