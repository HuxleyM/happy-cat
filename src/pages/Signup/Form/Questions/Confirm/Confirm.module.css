@import "../../../../../variables.css";

.header {
  padding: 36px 0px;
  border-bottom: 3pt solid var(--colour-grey);
  margin-bottom: 24px;
}

.header > h3 {
  font-weight: 700;
  margin-bottom: 5px;
  font-size: 20px;
}

.flexContainer {
  display: flex;
  flex-direction: column;
  flex-wrap: wrap;
  margin-bottom: 15px;
}

.space {
  padding-bottom: 16px;
  margin-bottom: 5px;
  border-bottom: 2pt solid var(--colour-grey);
}
.flexContainer > h4,
.bold {
  font-weight: 700;
}

.bold {
  font-size: 16px;
}

.flexContainer > div {
  margin: 10px 0px;
}
