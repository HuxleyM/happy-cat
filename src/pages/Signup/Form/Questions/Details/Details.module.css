.flexContainer {
  display: flex;
  flex-wrap: wrap;
  flex-direction: row;
  justify-content: space-between;
  width: 70vw;
  padding-top: 35px;
  /* background: blue; */
}
.questionWrapper {
  display: flex;
  flex-wrap: wrap;
  flex-direction: row;
  justify-content: space-between;
  width: 400px;
  margin-bottom: 30px;
}
.passwordInstructions {
  margin-top: 10px;
  font-weight: lighter;
  font-size: 14px;
  width: 100px;
  line-height: 16px;
}
