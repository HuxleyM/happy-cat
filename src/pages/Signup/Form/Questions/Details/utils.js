export const validEmail = () => {
    const email = document.getElementById("email").value;
    const emailRegex = /(?:[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*|"(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])*")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\[(?:(?:(2(5[0-5]|[0-4][0-9])|1[0-9][0-9]|[1-9]?[0-9]))\.){3}(?:(2(5[0-5]|[0-4][0-9])|1[0-9][0-9]|[1-9]?[0-9])|[a-z0-9-]*[a-z0-9]:(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21-\x5a\x53-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])+)\])/g;
    return !!email.match(emailRegex);
  };
  
  export const collectUserDetails = () => {
      const userName = document.getElementById('userName').value;
      const password = document.getElementById('password').value;
      const email = document.getElementById('email').value;
      return {userName, password, email}
  }
  
  export const validPassword = () => {
    const password = document.getElementById("password").value;
    const passwordRegex = /^(?=.*\d)(?=.*\W).{8,}$/g;
    return !!password.match(passwordRegex);
  };
  
  export const checkFieldsMatch = ({first, second}) => {
    const firstValue = document.getElementById(`${first}`).value;
    const secondValue = document.getElementById(`${second}`).value;
    return (firstValue === secondValue)
  };
  