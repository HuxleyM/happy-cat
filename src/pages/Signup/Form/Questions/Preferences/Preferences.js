import React, { useState, useContext, useRef } from "react";
import Styles from "./Preferences.module.css";
import { UserContext } from "../../../../../Context/userContext";

function errorsReducer(state, { key, error, message = "" }) {
  if (error) {
    state[key] = message;
  } else {
    delete state[key];
  }
  return state;
}

function Preferences({ handleFormSubmission }) {
  const { user, setUser } = useContext(UserContext);
  const [errors, setErrors] = useState({});
  const [answers, setAnswers] = useState("");

  const dogOptions = [
    { id: "0", text: "Unselected" },
    { id: "1", text: "Never" },
    { id: "2", text: "Some Times" },
    { id: "3", text: "Often" },
  ];

  const dogsAllowedField = useRef();
  const gifRateField = useRef();

  /// i d love to tidy this but do not know how to
  const isDisabled = () => {
    const errorAsArray = Object.keys(errors).length;
    const answersAsArray = Object.keys(answers).length;
    // disbaled needs false flag to disable and true to not
    const usable = errorAsArray === 0 && answersAsArray === 2 ? false : true;

    return (
      <button
        type="submit"
        disabled={usable}
        className={`${Styles.mainActionButton} `}
      >
        Save and next
      </button>
    );
  };

  const handleGifRateChange = () => {
    const reducerProps =
      gifRateField.current.value < 0 || gifRateField.current.value >= 10
        ? {
            error: true,
            key: "gifRate",
            message: "This is an invalid GIF rate",
          }
        : { error: false, key: "gifRate" };
    errorsReducer(errors, reducerProps);
    setAnswers({ ...answers, gifRateField: gifRateField.current.value });
  };

  const handleDogsAllowedChange = (e) => {
    const value = e.target.value;
    const reducerProps =
      value === "Unselected"
        ? { error: true, key: "dogsAllowed", message: "Please select option." }
        : { error: false, key: "dogsAllowed" };
    errorsReducer(errors, reducerProps);

    setAnswers({ ...answers, dogsAllowed: e.target.value });
  };

  const preferenceAnswers = () => {
    if (!answers.dogsAllowed) {
      errorsReducer(errors, {
        error: true,
        key: "dogsAllowed",
        message: "Select wether dogs are allowed",
      });
      // if errors break out and dont submit
      return;
    }
    const { gifRate, dogsAllowed } = answers;
    setUser({ ...user, gifRate, dogsAllowed });
  };
  return (
    <form
      onSubmit={(event) => {
        event.preventDefault();
        handleFormSubmission(preferenceAnswers);
      }}
    >
      <div className={Styles.flexContainer}>
        <div>
          <div className={Styles.questionWrapper}>
            <label htmlFor="dogsAllowed">Dogs allowed:</label>
            <div>
              <select
                id="dogsAllowed"
                name="dogsAllowed"
                ref={dogsAllowedField}
                onChange={handleDogsAllowedChange}
                value={user.dogsAllowed}
                required
              >
                {dogOptions.map((option) => (
                  <option key={option.id} value={option.text}>
                    {option.text}
                  </option>
                ))}
              </select>
              {errors.dogsAllowed && (
                <div className={Styles.inputError}>{errors.dogsAllowed}</div>
              )}
            </div>
          </div>

          <div className={Styles.questionWrapper}>
            <label htmlFor="gifRate">No. GIF's a day:</label>
            <div>
              <input
                min="1"
                max="10"
                type="number"
                id="gifRate"
                name="gifRate"
                required
                ref={gifRateField}
                value={user.gifRate}
                onChange={handleGifRateChange}
              ></input>
              {errors.gifRate && (
                <div className={Styles.inputError}>{errors.gifRate}</div>
              )}
            </div>
          </div>
        </div>
      </div>
      <div className={Styles.moveToEnd}>{isDisabled()}</div>
    </form>
  );
}

export default Preferences;
