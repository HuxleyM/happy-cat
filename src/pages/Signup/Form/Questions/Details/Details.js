import React, { useState, useContext, useRef, useEffect } from "react";
import Styles from "./Details.module.css";
import * as utils from "./utils";
import { UserContext } from "../../../../../Context/userContext";
import {FontAwesomeIcon} from '@fortawesome/react-fontawesome'
import { faEye, faEyeSlash } from "@fortawesome/free-solid-svg-icons";

function Details({ handleFormSubmission }) {
  const { user, setUser } = useContext(UserContext);
  const [errors, setErrors] = useState({});

  const [userName, setUserName] = useState("");
  const [password, setPassword] = useState("");
  const [passwordRetype, setPasswordRetype] = useState("");
  const [email, setEmail] = useState("");
  const [emailRetype, setEmailRetype] = useState("");
  const [passwordShown, setPasswordShown] = useState(false);

  const userNameField = useRef();
  const passwordField = useRef();
  const passwordRetypeField = useRef();
  const emailField = useRef();
  const emailRetypeField = useRef();

  const handleUserNameChange = () => {
    setUserName(userNameField.current.value);
  };

  const isDisabled = () => {
    const errorAsArray = Object.keys(errors).length;

    if (
      userName &&
      password &&
      passwordRetype &&
      emailRetype &&
      email &&
      !errorAsArray
    ) {
      return (
        <button type="submit" className={`${Styles.mainActionButton} `}>
          Save and next
        </button>
      );
    }
    return (
      <button type="submit" disabled className={`${Styles.mainActionButton} `}>
        Save and next
      </button>
    );
  };

  const ifExistingErrorsDelete = ({ key }) => {
    if (errors[key]) {
      const newErrors = { ...errors };
      delete newErrors[key];
      setErrors({ ...newErrors });
    }
  };

  const handleEmailChange = () => {
    if (!utils.validEmail(emailField.current.value)) {
      setErrors({ ...errors, email: "This is not a valid email." });
    } else {
      setEmail(emailField.current.value);
      ifExistingErrorsDelete({ key: "email" });
    }
  };

  const handleEmailRetypeChange = () => {
    if (emailField.current.value !== emailRetypeField.current.value) {
      setErrors({ ...errors, emailRetype: "Emails do not match." });
    } else {
      setEmailRetype(emailRetypeField.current.value);
      ifExistingErrorsDelete({ key: "emailRetype" });
    }
  };

  const handlePasswordChange = () => {
    if (!utils.validPassword(passwordField.current.value)) {
      setErrors({ ...errors, password: "This is not a valid password." });
    } else {
      setPassword(passwordField.current.value);
      ifExistingErrorsDelete({ key: "password" });
    }
  };

  const handlePasswordRetypeChange = () => {
    if (passwordField.current.value !== passwordRetypeField.current.value) {
      setErrors({ ...errors, passwordRetype: "Password do not match." });
    } else {
      setPasswordRetype(passwordRetypeField.current.value);
      ifExistingErrorsDelete({ key: "passwordRetype" });
    }
  };

  const detailsAnswers = () => {
    setUser({ ...user, userName, email, password });
  };

  const togglePasswordVisiblity = () => {
    console.log('hello')
    setPasswordShown(passwordShown ? false : true);
  };

  return (
    <form
      onSubmit={(event) => {
        event.preventDefault();
        handleFormSubmission(detailsAnswers);
      }}
    >
      <div className={Styles.flexContainer}>
        <div>
          <div className={Styles.questionWrapper}>
            <label htmlFor="userName">User name:</label>
            <div>
              <input
                type="text"
                id="userName"
                name="userName"
                ref={userNameField}
                onChange={handleUserNameChange}
                value={user.userName}
                required
              ></input>
              {errors.username && (
                <div className={Styles.inputError}>emails do not match</div>
              )}
            </div>
          </div>

          <div className={Styles.questionWrapper}>
            <label htmlFor="email">Email:</label>
            <div>
              <input
                type="email"
                id="email"
                name="email"
                required
                ref={emailField}
                value={user.email}
                onChange={handleEmailChange}
              ></input>
              {errors.email && (
                <div className={Styles.inputError}>{errors.email}</div>
              )}
            </div>
          </div>

          <div className={Styles.questionWrapper}>
            <label htmlFor="emailRetype">Retype email:</label>
            <div>
              <input
                type="email"
                id="emailRetype"
                name="emailRetype"
                ref={emailRetypeField}
                value={user.email}
                onChange={handleEmailRetypeChange}
                required
              ></input>
              {errors.emailRetype && (
                <div className={Styles.inputError}>{errors.emailRetype}</div>
              )}
            </div>
          </div>
        </div>

        <div>
          <div className={Styles.questionWrapper}>
            <div>
              <label htmlFor="password">Password:</label>
              <span className={Styles.passwordVisibility} onClick={()=> togglePasswordVisiblity()}>
                <FontAwesomeIcon icon={passwordShown ? faEye : faEyeSlash} />
              </span>
              <p className={Styles.passwordInstructions}>
                Password must contain at least 8 digits, a special character and
                a number.
              </p>
            </div>
            <div>
              <input
                type={passwordShown ? "text" : "password"}
                id="password"
                name="password"
                ref={passwordField}
                value={user.password}
                onChange={handlePasswordChange}
                required
              ></input>
              {errors.password && (
                <div className={Styles.inputError}>{errors.password}</div>
              )}
            </div>
          </div>

          <div className={Styles.questionWrapper}>
            <label htmlFor="passwordRetype">Retype password:</label>
            <div>
              <input
                type={passwordShown ? "text" : "password"}
                id="passwordRetype"
                name="passwordRetype"
                ref={passwordRetypeField}
                value={user.password}
                onChange={handlePasswordRetypeChange}
                required
              ></input>
              <div>
                {errors.passwordRetype && (
                  <div className={Styles.inputError}>
                    {errors.passwordRetype}
                  </div>
                )}
              </div>
            </div>
          </div>
        </div>
      </div>
      <div className={Styles.moveToEnd}>
        {isDisabled()}
      </div>
    </form>
  );
}

export default Details;
