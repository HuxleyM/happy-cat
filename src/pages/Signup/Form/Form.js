import React, { useContext, useState } from "react";
import Styles from "./Form.module.css";
import { UserContext } from "../../../Context/userContext";
import Details from "./Questions/Details/Details";
import Preferences from "./Questions/Preferences/Preferences";
import Confirm from "./Questions/Confirm/Confirm";



function Form() {
  const { user, setUser } = useContext(UserContext);
  const [formProgress, setFormProgress] = useState({
    questionsAnswered: 0,
    currentlyOnQuestion: 0,
  });



const movePage = (formProgress, setFormProgress, index) => {
    setTimeout(() => {
      let newPage = (formProgress.currentlyOnQuestion += index);
      setFormProgress({ ...formProgress, currentlyOnQuestion: newPage });
    }, 200);
  };


  const questionToShow = ({ currentlyOnQuestion }) => {
    if (!currentlyOnQuestion && currentlyOnQuestion !== 0)
      return <div>Loading</div>;
        const Questions = [
        <Details
            formProgress={formProgress}
            setFormProgress={setFormProgress}
        />,
        <Preferences
            formProgress={formProgress}
            setFormProgress={setFormProgress}
        />,
        <Confirm
            formProgress={formProgress}
            setFormProgress={setFormProgress}
        />,
        ];
        return Questions[currentlyOnQuestion];
    };
  const question = questionToShow(formProgress);
  return (
    <div>
      <div className={Styles.addBreathingSpace}>
        <div className={Styles.headerWrapper}>
          <h2 className={Styles.header}>SignUp</h2>
          <nav className={Styles.navBar}>
            <ul>
              <li>Details</li>
              <li>Preferences</li>
              <li>Confirm</li>
            </ul>
          </nav>
        </div>
        <div>
          <p>
            Man-of-war lateen sail spirits gunwalls gibbet tackle reef sails
            scuppers hands splice the main brace. Man-of-war plunder Davy Jones'
            Locker Buccaneer smartly keelhaul dance the hempen jig yo-ho-ho
            Plate Fleet list. Handsomely Pieces of Eight cable no prey, no pay
            lanyard Letter of Marque quarterdeck gangplank rum lookout. Aye log
            cutlass matey execution dock pressgang Shiver me timbers grog
            blossom jury mast pinnace. Cutlass hail-shot tender crimp Corsair me
            fore ye splice the main brace chandler. Nipperkin execution dock
            bowsprit smartly bring a spring upon her cable gally tack fore brig
            spirits.
          </p>
        </div>
      </div>
      <div className={Styles.questionContainer}>
        {formProgress.questionsAnswered >= formProgress.currentlyOnQuestion &&
          formProgress.currentlyOnQuestion > 0 && (
            <div
              className={Styles.leftArrow}
              onClick={() => movePage(formProgress, setFormProgress, -1)}
            >
              {" "}
              back{" "}
            </div>
          )}
        {question}
        {formProgress.questionsAnswered > formProgress.currentlyOnQuestion &&
          formProgress.currentlyOnQuestion < 3 && (
            <div
              className={Styles.rightArrow}
              onClick={() => movePage(formProgress, setFormProgress, 1)}
            >
              {" "}
              next{" "}
            </div>
          )}
      </div>
    </div>
  );
}

export default Form;
