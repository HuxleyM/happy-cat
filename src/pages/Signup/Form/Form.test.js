import React from 'react';
import { mount } from 'enzyme';
import Form from './Form';
import Details from './Questions/Details/Details';
import Preferences from './Questions/Preferences/Preferences';
import Confirm from './Questions/Confirm/Confirm'
import { UserContext } from '../../../Context/userContext';

describe('Form component', () => {

    it('Should color the nav bar in respect to a users position in the questions', () => {

    })

    it('If a user is on question 1 render Details', ()=>{
        const context = {user:{ currentlyOnQuestion:0}};
        const wrapper = mount(
          <UserContext.Provider value={context}>
            <Form />
          </UserContext.Provider>,
        );
        expect(wrapper.containsMatchingElement(<Details />)).toEqual(true);
        expect(wrapper.containsMatchingElement(<Preferences />)).toEqual(false);
        expect(wrapper.containsMatchingElement(<Confirm />)).toEqual(false);
    })
    it('If a user is on question 2 render Preferences', ()=>{
        const context = {user:{ currentlyOnQuestion:1}};
        const wrapper = mount(
          <UserContext.Provider value={context}>
            <Form />
          </UserContext.Provider>,
        );
        expect(wrapper.containsMatchingElement(<Details />)).toEqual(false);
        expect(wrapper.containsMatchingElement(<Preferences />)).toEqual(true);
        expect(wrapper.containsMatchingElement(<Confirm />)).toEqual(false);
    })
    it('If a user is on question 3 render Confirm', ()=>{
        const context = {user:{ currentlyOnQuestion:2}};
        const wrapper = mount(
          <UserContext.Provider value={context}>
            <Form />
          </UserContext.Provider>,
        );
        expect(wrapper.containsMatchingElement(<Details />)).toEqual(false);
        expect(wrapper.containsMatchingElement(<Preferences />)).toEqual(false);
        expect(wrapper.containsMatchingElement(<Confirm />)).toEqual(true);
    })

    describe('Navagation', ()=>{
        it('if i have answered all questions i should be able to navagate back and forth', ()=> {

        })
    })
})