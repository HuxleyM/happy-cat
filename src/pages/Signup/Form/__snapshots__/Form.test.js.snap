// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component Navigation if i have answered all questions i should be able to navagate back and forth 1`] = `
<Form>
  <div>
    <div
      className="addBreathingSpace"
    >
      <div
        className="headerWrapper"
      >
        <h2
          className="header"
        >
          SignUp
        </h2>
        <nav
          className="navBar"
        >
          <ul>
            <li>
              Details
            </li>
            <li>
              Preferences
            </li>
            <li>
              Confirm
            </li>
          </ul>
        </nav>
      </div>
      <div>
        <p>
          Man-of-war lateen sail spirits gunwalls gibbet tackle reef sails scuppers hands splice the main brace. Man-of-war plunder Davy Jones' Locker Buccaneer smartly keelhaul dance the hempen jig yo-ho-ho Plate Fleet list. Handsomely Pieces of Eight cable no prey, no pay lanyard Letter of Marque quarterdeck gangplank rum lookout. Aye log cutlass matey execution dock pressgang Shiver me timbers grog blossom jury mast pinnace. Cutlass hail-shot tender crimp Corsair me fore ye splice the main brace chandler. Nipperkin execution dock bowsprit smartly bring a spring upon her cable gally tack fore brig spirits.
        </p>
      </div>
    </div>
    <div
      className="questionContainer"
    >
      <div
        className="leftArrow"
        onClick={[Function]}
      >
         back 
      </div>
    </div>
  </div>
</Form>
`;
